compiler/rustc_middle/src/ty/context.rs

```diff
@@ -2439,6 +2439,14 @@ impl<'tcx> TyCtxt<'tcx> {
     pub fn trait_solver_next(self) -> bool {
         self.sess.opts.unstable_opts.trait_solver == rustc_session::config::TraitSolver::Next
     }
+
+    pub fn lower_impl_trait_in_trait_to_assoc_ty(self) -> bool {
+        self.sess.opts.unstable_opts.lower_impl_trait_in_trait_to_assoc_ty
+    }
+
+    pub fn use_new_lowering_impl_trait_in_trait_strategy(self, def_id: DefId) -> bool {
+        self.lower_impl_trait_in_trait_to_assoc_ty() && self.opt_rpitit_info(def_id).is_some()
+    }
```

Comment:
```suggestion
    pub fn is_impl_trait_in_trait(self, def_id: DefId) -> bool {
        if self.lower_impl_trait_in_trait_to_assoc_ty() {
            self.def_kind(def_id) == DefKind::AssocTy && self.opt_rpitit_info(def_id).is_some()
        } else {
            self.def_kind(def_id) == DefKind::ImplTraitPlaceholder
    }
```

Reply:
Then, at least for *most* of the callsites that look like:

```
if ty::Alias(ty::Projection, alias_ty) = ty.kind() && tcx.def_kind(alias_ty.def_id) == DefKind::ImplTraitPlaceholder
```

We can just do:

```
if ty::Alias(ty::Projection, alias_ty) = ty.kind() && tcx. is_impl_trait_in_trait(alias_ty.def_id)
```

Reply:
Isn't `self.def_kind(def_id) == DefKind::AssocTy` redundant?

Reply:
I think for now, but ideally I'd probably want `opt_rpitit_info` to ICE for non-associated-types, so we don't call it for random other def ids that aren't ever associated with RPITITs.

That's probably a refactoring I'll do on the side, though, once you're done with the bulk of this refactoring work.